src/hooks/useControllableState.ts

```diff
@@ -0,0 +1,113 @@
+import React from 'react'
+declare let __DEV__: boolean
+
+type ControllableStateOptions<T> = {
+  /**
+   * A unique name for the state value.
+   */
+  name?: string
+  /**
+   * The default value used for the state. This will be
+   * the fallback value used if `value` is not defined.
+   * */
+  defaultValue: T
+  /**
+   * A controlled value. Omitting this means that the state is uncontrolled.
+   */
+  value?: T
+  /**
+   * An optional function that is called when the value of the state changes.
+   * This is useful for communicating to parents of controlled components
+   * that the value is requesting to be changed.
+   */
+  onChange?: (value: T) => void
+}
+
+/**
+ * This custom hook simplifies the behavior of a component if it has state that
+ * can be both controlled and uncontrolled. It functions identical to a
+ * useState() hook and provides [state, setState] for you to use. You can use
+ * the `onChange` argument to allow updates to the `state` to be communicated to
+ * owners of controlled components.
+ *
+ * Note: This hook will warn if a component is switching from controlled to
+ * uncontrolled, or vice-versa.
+ */
+export function useControllableState<T>({
+  name = 'custom',
+  defaultValue,
+  value,
+  onChange
+}: ControllableStateOptions<T>): [T, React.Dispatch<React.SetStateAction<T>>] {
+  const [state, internalSetState] = React.useState(value ?? defaultValue)
+  const controlled = React.useRef<boolean | null>(null)
+
+  if (controlled.current === null) {
+    controlled.current = value !== undefined
+  }
+
+  function setState(stateOrUpdater: T | ((prevState: T) => T)) {
```

Comment:
Idea from the stable state setter comment, this might be able to become:

```tsx
const savedOnChange = React.useRef(onChange);

React.useEffect(() => {
  savedOnChange.current = onChange;
});

const setState = React.useCallback((stateOrUpdater: T | ((prevState: T) => T)) => {
  const value =
    typeof stateOrUpdater === 'function'
      ? // @ts-ignore stateOrUpdater is a function
        stateOrUpdater(state)
      : stateOrUpdater

  if (controlled.current === false) {
    internalSetState(value)
  }

  if (onChange) {
    savedOnChange.current?.(value)
  }
}, [state]);

// --snip--

return [state, setState];
```

It'd be great to also get rid of `state` from the dep array but I didn't see a quick way to do so 🤔

Reply:
This is great feedback. Thank you! Is this what you had in mind? https://github.com/primer/react/pull/2373/commits/eeaf486830f268bc94bab2257f629be8e7468b19